File: sakinah/js/core/qibla.js
/**
 * اتجاه القبلة بدقة عالية
 * ----------------------
 * 1) الدائرة العظمى على كرة (الصيغة الكلاسيكية) — كافية لمعظم الاستخدامات.
 * 2) الجيوديسية على مجسّم WGS-84 بطريقة Vincenty (1975) — أدق بفارق قد يبلغ 0.2° في بعض المناطق،
 *    وتعطي المسافة الحقيقية إلى الكعبة.
 * 3) التحقق بالشمس: لحظات يكون فيها سمت الشمس مساويًا لاتجاه القبلة (أو معاكسًا له فيشير الظل إلى القبلة)،
 *    وحادثتا تعامد الشمس على الكعبة سنويًا — طرق مستقلة عن البوصلة المغناطيسية.
 */
import { d2r, r2d, unwindAngle, quadrantShiftAngle, sunPosition, SolarTime, julianDay, solarCoordinates } from './astro.js';

/** إحداثيات الكعبة المشرفة (مركز البناء) */
export const KAABA = { latitude: 21.4225241, longitude: 39.8261818 };

// WGS-84
const A = 6378137.0, F = 1 / 298.257223563, B = A * (1 - F);

/** اتجاه القبلة (دائرة عظمى) بالدرجات من الشمال الحقيقي باتجاه عقارب الساعة */
export function qiblaSpherical(latitude, longitude) {
  const φ1 = d2r(latitude), φ2 = d2r(KAABA.latitude), Δλ = d2r(KAABA.longitude - longitude);
  const y = Math.sin(Δλ);
  const x = Math.cos(φ1) * Math.tan(φ2) - Math.sin(φ1) * Math.cos(Δλ);
  return unwindAngle(r2d(Math.atan2(y, x)));
}

/** المسافة على الكرة (هافرساين) بالكيلومترات */
export function distanceSphericalKm(lat1, lon1, lat2 = KAABA.latitude, lon2 = KAABA.longitude) {
  const R = 6371.0088;
  const φ1 = d2r(lat1), φ2 = d2r(lat2), dφ = d2r(lat2 - lat1), dλ = d2r(lon2 - lon1);
  const a = Math.sin(dφ / 2) ** 2 + Math.cos(φ1) * Math.cos(φ2) * Math.sin(dλ / 2) ** 2;
  return 2 * R * Math.asin(Math.min(1, Math.sqrt(a)));
}

/**
 * حلّ Vincenty العكسي على مجسّم WGS-84.
 * @returns {{distanceKm:number, initialBearing:number, finalBearing:number, converged:boolean}}
 */
export function vincentyInverse(lat1, lon1, lat2 = KAABA.latitude, lon2 = KAABA.longitude) {
  const φ1 = d2r(lat1), φ2 = d2r(lat2);
  const L = d2r(lon2 - lon1);
  const tanU1 = (1 - F) * Math.tan(φ1), cosU1 = 1 / Math.sqrt(1 + tanU1 * tanU1), sinU1 = tanU1 * cosU1;
  const tanU2 = (1 - F) * Math.tan(φ2), cosU2 = 1 / Math.sqrt(1 + tanU2 * tanU2), sinU2 = tanU2 * cosU2;
  let λ = L, λʹ, iterations = 0;
  let sinλ, cosλ, sinσ, cosσ, σ, sinα, cos2α, cos2σm, C;
  do {
    sinλ = Math.sin(λ); cosλ = Math.cos(λ);
    const sinSqσ = (cosU2 * sinλ) ** 2 + (cosU1 * sinU2 - sinU1 * cosU2 * cosλ) ** 2;
    sinσ = Math.sqrt(sinSqσ);
    if (sinσ === 0) return { distanceKm: 0, initialBearing: 0, finalBearing: 0, converged: true }; // نقطتان متطابقتان
    cosσ = sinU1 * sinU2 + cosU1 * cosU2 * cosλ;
    σ = Math.atan2(sinσ, cosσ);
    sinα = (cosU1 * cosU2 * sinλ) / sinσ;
    cos2α = 1 - sinα * sinα;
    cos2σm = cos2α !== 0 ? cosσ - (2 * sinU1 * sinU2) / cos2α : 0; // خط الاستواء
    C = (F / 16) * cos2α * (4 + F * (4 - 3 * cos2α));
    λʹ = λ;
    λ = L + (1 - C) * F * sinα * (σ + C * sinσ * (cos2σm + C * cosσ * (-1 + 2 * cos2σm * cos2σm)));
  } while (Math.abs(λ - λʹ) > 1e-12 && ++iterations < 200);
  const converged = iterations < 200;
  if (!converged) {
    // نقاط شبه متقابلة قطريًا: نعود إلى الحل الكروي
    return { distanceKm: distanceSphericalKm(lat1, lon1, lat2, lon2), initialBearing: qiblaSpherical(lat1, lon1), finalBearing: NaN, converged: false };
  }
  const uSq = (cos2α * (A * A - B * B)) / (B * B);
  const Acoef = 1 + (uSq / 16384) * (4096 + uSq * (-768 + uSq * (320 - 175 * uSq)));
  const Bcoef = (uSq / 1024) * (256 + uSq * (-128 + uSq * (74 - 47 * uSq)));
  const Δσ = Bcoef * sinσ * (cos2σm + (Bcoef / 4) * (cosσ * (-1 + 2 * cos2σm * cos2σm) - (Bcoef / 6) * cos2σm * (-3 + 4 * sinσ * sinσ) * (-3 + 4 * cos2σm * cos2σm)));
  const s = B * Acoef * (σ - Δσ);
  const α1 = Math.atan2(cosU2 * sinλ, cosU1 * sinU2 - sinU1 * cosU2 * cosλ);
  const α2 = Math.atan2(cosU1 * sinλ, -sinU1 * cosU2 + cosU1 * sinU2 * cosλ);
  return { distanceKm: s / 1000, initialBearing: unwindAngle(r2d(α1)), finalBearing: unwindAngle(r2d(α2)), converged: true };
}

/**
 * معلومات القبلة الكاملة لموقع.
 * bearing: الاتجاه الجيوديسي (الأدق)؛ bearingSpherical: الحل الكروي؛ distanceKm: المسافة الجيوديسية
 */
export function qiblaInfo(latitude, longitude) {
  const sph = qiblaSpherical(latitude, longitude);
  const v = vincentyInverse(latitude, longitude);
  const bearing = v.converged ? v.initialBearing : sph;
  return {
    bearing, bearingSpherical: sph, distanceKm: v.distanceKm,
    difference: quadrantShiftAngle(bearing - sph),
    compassPoint: compassPointAr(bearing),
    /** قرب النقطة المقابلة للكعبة (شرق بولينيزيا) لا يتقارب حل Vincenty والاتجاه غير محدد عمليًا */
    antipodal: !v.converged || v.distanceKm > 19900,
  };
}

const POINTS_AR = ['شمال', 'شمال شرق', 'شرق', 'جنوب شرق', 'جنوب', 'جنوب غرب', 'غرب', 'شمال غرب'];
export function compassPointAr(bearing) {
  return POINTS_AR[Math.round(unwindAngle(bearing) / 45) % 8];
}

/** الفرق الموقّع بين اتجاهين (-180..180]: موجب = القبلة على يمين المرجع */
export function signedDifference(target, reference) {
  return quadrantShiftAngle(unwindAngle(target - reference));
}

/**
 * لحظات التحقق بالشمس خلال يوم مدني في منطقة زمنية:
 * - sunAtQibla: الشمس في اتجاه القبلة تمامًا (استقبل الشمس = استقبلت القبلة)
 * - shadowAtQibla: الشمس معاكسة للقبلة (ظلّ العمود الرأسي يشير إلى القبلة)
 * @returns {{sunAtQibla: Date|null, shadowAtQibla: Date|null}}
 */
export function sunQiblaMoments(latitude, longitude, civil, tz) {
  const bearing = qiblaInfo(latitude, longitude).bearing;
  const start = localMidnightUTC(civil, tz);
  const find = (target) => {
    let prev = null, prevT = null;
    for (let m = 0; m <= 1440; m += 2) {
      const t = new Date(start.getTime() + m * 60000);
      const p = sunPosition(t, latitude, longitude);
      const f = signedDifference(p.azimuth, target);
      if (prev !== null && Math.sign(f) !== Math.sign(prev) && Math.abs(f - prev) < 180) {
        // تنقيح بالتنصيف
        let lo = prevT, hi = t, flo = prev;
        for (let i = 0; i < 25; i++) {
          const mid = new Date((lo.getTime() + hi.getTime()) / 2);
          const fm = signedDifference(sunPosition(mid, latitude, longitude).azimuth, target);
          if (Math.sign(fm) === Math.sign(flo)) { lo = mid; flo = fm; } else hi = mid;
        }
        const res = new Date(Math.round((lo.getTime() + hi.getTime()) / 2000) * 1000);
        const pr = sunPosition(res, latitude, longitude);
        // نرفض العبور الزائف قرب سمت الرأس (السمت ينقلب 180° في ثوانٍ) ونتأكد أن الفرق المتبقي صغير فعلًا
        if (pr.altitude > 0 && pr.altitude < 85 && Math.abs(signedDifference(pr.azimuth, target)) < 0.5) return res;
      }
      prev = f; prevT = t;
    }
    return null;
  };
  return { bearing, sunAtQibla: find(bearing), shadowAtQibla: find(unwindAngle(bearing + 180)) };
}

/** منتصف الليل المحلي (بداية اليوم المدني) كلحظة UTC */
export function localMidnightUTC(civil, tz) {
  const fmt = new Intl.DateTimeFormat('en-US', { timeZone: tz, hourCycle: 'h23', year: 'numeric', month: 'numeric', day: 'numeric', hour: 'numeric', minute: 'numeric' });
  const wall = (ms) => { const p = fmt.formatToParts(new Date(ms)).reduce((o, x) => (o[x.type] = x.value, o), {}); return Date.UTC(+p.year, +p.month - 1, +p.day, +p.hour, +p.minute); };
  const target = Date.UTC(civil.year, civil.month - 1, civil.day, 0, 0, 0);
  // نبدأ من تخمين UTC ثم نصحّح بفارق المنطقة (تكرار حتى الاستقرار)
  let guess = target;
  for (let i = 0; i < 4; i++) { const next = target - (wall(guess) - guess); if (next === guess) break; guess = next; }
  // إن لم تكن الساعة 00:00 موجودة (تحويل التوقيت الصيفي عند منتصف الليل كما في القاهرة) نأخذ أول لحظة في اليوم المدني المطلوب
  if (wall(guess) < target) { let t = guess; for (let i = 0; i < 180 && wall(t) < target; i++) t += 60000; guess = t; }
  return new Date(guess);
}

/**
 * حادثتا تعامد الشمس على الكعبة في سنة معيّنة (نحو 27/28 مايو و15/16 يوليو، الساعة ~09:18 و~09:27 UTC).
 * @returns {Array<{time: Date, altitude: number}>}
 */
export function kaabaZenithEvents(year) {
  const events = [];
  for (const [m1, d1, m2, d2] of [[5, 20, 6, 5], [7, 8, 7, 24]]) {
    let best = null;
    for (let jd = julianDay(year, m1, d1); jd <= julianDay(year, m2, d2); jd += 1) {
      const civil = jdToCivil(jd);
      const st = new SolarTime(civil, KAABA);
      const t = new Date(Date.UTC(civil.year, civil.month - 1, civil.day) + st.transit * 3600000);
      const alt = sunPosition(t, KAABA.latitude, KAABA.longitude).altitude;
      if (!best || alt > best.altitude) best = { time: t, altitude: alt };
    }
    events.push(best);
  }
  return events;
}

function jdToCivil(jd) {
  const d = new Date((jd - 2440587.5) * 86400000);
  return { year: d.getUTCFullYear(), month: d.getUTCMonth() + 1, day: d.getUTCDate() };
}

export { sunPosition, solarCoordinates };

/**
 * نقاط على قوس الدائرة العظمى (أقصر مسار على الكرة) من نقطة إلى أخرى — للرسم على الخريطة.
 * استيفاء كروي (slerp) على المتجهات الواحدة؛ يعيد [[lat, lon], ...] بعدد n+1 نقطة.
 */
export function greatCirclePoints(lat1, lon1, lat2, lon2, n = 64) {
  const d = Math.PI / 180;
  const toVec = (la, lo) => [Math.cos(la * d) * Math.cos(lo * d), Math.cos(la * d) * Math.sin(lo * d), Math.sin(la * d)];
  const a = toVec(lat1, lon1), b = toVec(lat2, lon2);
  const dot = Math.max(-1, Math.min(1, a[0] * b[0] + a[1] * b[1] + a[2] * b[2]));
  const omega = Math.acos(dot); const pts = [];
  if (omega < 1e-7) return [[lat1, lon1], [lat2, lon2]]; // النقطتان متطابقتان عمليًا
  for (let i = 0; i <= n; i++) {
    const t = i / n; const s1 = Math.sin((1 - t) * omega) / Math.sin(omega), s2 = Math.sin(t * omega) / Math.sin(omega);
    const x = s1 * a[0] + s2 * b[0], y = s1 * a[1] + s2 * b[1], z = s1 * a[2] + s2 * b[2];
    pts.push([Math.atan2(z, Math.hypot(x, y)) / d, Math.atan2(y, x) / d]);
  }
  return pts;
}

File: sakinah-native/Sources/SakinahCore/Qibla.swift
import Foundation

/// اتجاه القبلة: دائرة عظمى، وجيوديسية Vincenty على WGS-84، والتحقق بالشمس — مطابق لـ qibla.js
public enum Qibla {
  /// إحداثيات الكعبة المشرفة (مركز البناء)
  public static let kaaba = Coordinates(latitude: 21.4225241, longitude: 39.8261818)
  static let A = 6378137.0, F = 1 / 298.257223563
  static let B = A * (1 - F)

  /// اتجاه القبلة (دائرة عظمى) بالدرجات من الشمال الحقيقي باتجاه عقارب الساعة
  public static func spherical(latitude: Double, longitude: Double) -> Double {
    let φ1 = d2r(latitude), φ2 = d2r(kaaba.latitude), Δλ = d2r(kaaba.longitude - longitude)
    let y = sin(Δλ)
    let x = cos(φ1) * tan(φ2) - sin(φ1) * cos(Δλ)
    return unwindAngle(r2d(atan2(y, x)))
  }
  /// المسافة على الكرة (هافرساين) بالكيلومترات
  public static func distanceSphericalKm(lat1: Double, lon1: Double, lat2: Double = kaaba.latitude, lon2: Double = kaaba.longitude) -> Double {
    let R = 6371.0088
    let φ1 = d2r(lat1), φ2 = d2r(lat2), dφ = d2r(lat2 - lat1), dλ = d2r(lon2 - lon1)
    let a = pow(sin(dφ / 2), 2) + cos(φ1) * cos(φ2) * pow(sin(dλ / 2), 2)
    return 2 * R * asin(min(1, sqrt(a)))
  }

  public struct Vincenty: Sendable { public let distanceKm: Double; public let initialBearing: Double; public let finalBearing: Double; public let converged: Bool }
  /// حلّ Vincenty العكسي على مجسّم WGS-84
  public static func vincentyInverse(lat1: Double, lon1: Double, lat2: Double = kaaba.latitude, lon2: Double = kaaba.longitude) -> Vincenty {
    let φ1 = d2r(lat1), φ2 = d2r(lat2)
    let L = d2r(lon2 - lon1)
    let tanU1 = (1 - F) * tan(φ1), cosU1 = 1 / sqrt(1 + tanU1 * tanU1), sinU1 = tanU1 * cosU1
    let tanU2 = (1 - F) * tan(φ2), cosU2 = 1 / sqrt(1 + tanU2 * tanU2), sinU2 = tanU2 * cosU2
    var λ = L, λʹ = 0.0, iterations = 0
    var sinλ = 0.0, cosλ = 0.0, sinσ = 0.0, cosσ = 0.0, σ = 0.0, sinα = 0.0, cos2α = 0.0, cos2σm = 0.0, C = 0.0
    repeat {
      sinλ = sin(λ); cosλ = cos(λ)
      let sinSqσ = pow(cosU2 * sinλ, 2) + pow(cosU1 * sinU2 - sinU1 * cosU2 * cosλ, 2)
      sinσ = sqrt(sinSqσ)
      if sinσ == 0 { return Vincenty(distanceKm: 0, initialBearing: 0, finalBearing: 0, converged: true) }
      cosσ = sinU1 * sinU2 + cosU1 * cosU2 * cosλ
      σ = atan2(sinσ, cosσ)
      sinα = (cosU1 * cosU2 * sinλ) / sinσ
      cos2α = 1 - sinα * sinα
      cos2σm = cos2α != 0 ? cosσ - (2 * sinU1 * sinU2) / cos2α : 0
      C = (F / 16) * cos2α * (4 + F * (4 - 3 * cos2α))
      λʹ = λ
      λ = L + (1 - C) * F * sinα * (σ + C * sinσ * (cos2σm + C * cosσ * (-1 + 2 * cos2σm * cos2σm)))
      iterations += 1
    } while abs(λ - λʹ) > 1e-12 && iterations < 200
    let converged = iterations < 200
    if !converged {
      return Vincenty(distanceKm: distanceSphericalKm(lat1: lat1, lon1: lon1, lat2: lat2, lon2: lon2), initialBearing: spherical(latitude: lat1, longitude: lon1), finalBearing: .nan, converged: false)
    }
    let uSq = (cos2α * (A * A - B * B)) / (B * B)
    let Acoef = 1 + (uSq / 16384) * (4096 + uSq * (-768 + uSq * (320 - 175 * uSq)))
    let Bcoef = (uSq / 1024) * (256 + uSq * (-128 + uSq * (74 - 47 * uSq)))
    let Δσ = Bcoef * sinσ * (cos2σm + (Bcoef / 4) * (cosσ * (-1 + 2 * cos2σm * cos2σm) - (Bcoef / 6) * cos2σm * (-3 + 4 * sinσ * sinσ) * (-3 + 4 * cos2σm * cos2σm)))
    let s = B * Acoef * (σ - Δσ)
    let α1 = atan2(cosU2 * sinλ, cosU1 * sinU2 - sinU1 * cosU2 * cosλ)
    let α2 = atan2(cosU1 * sinλ, -sinU1 * cosU2 + cosU1 * sinU2 * cosλ)
    return Vincenty(distanceKm: s / 1000, initialBearing: unwindAngle(r2d(α1)), finalBearing: unwindAngle(r2d(α2)), converged: true)
  }

  public struct Info: Sendable {
    /// الاتجاه الجيوديسي (الأدق)
    public let bearing: Double
    public let bearingSpherical: Double
    public let distanceKm: Double
    public let difference: Double
    public let compassPoint: String
    /// قرب النقطة المقابلة للكعبة: الاتجاه غير محدد عمليًا
    public let antipodal: Bool
  }
  /// معلومات القبلة الكاملة لموقع
  public static func info(latitude: Double, longitude: Double) -> Info {
    let sph = spherical(latitude: latitude, longitude: longitude)
    let v = vincentyInverse(lat1: latitude, lon1: longitude)
    let bearing = v.converged ? v.initialBearing : sph
    return Info(bearing: bearing, bearingSpherical: sph, distanceKm: v.distanceKm, difference: quadrantShiftAngle(bearing - sph),
                compassPoint: compassPointAr(bearing), antipodal: !v.converged || v.distanceKm > 19900)
  }
  /// هامش خطأ الاتجاه (درجات) الناتج عن عدم يقين الموقع
  public static func bearingUncertainty(distanceKm: Double, locationErrorM: Double) -> Double {
    let d = distanceKm * 1000, e = max(0, locationErrorM)
    if !(d > 0) || e >= d { return 180 }
    return r2d(asin(e / d))
  }
  static let pointsAr = ["شمال", "شمال شرق", "شرق", "جنوب شرق", "جنوب", "جنوب غرب", "غرب", "شمال غرب"]
  public static func compassPointAr(_ bearing: Double) -> String { pointsAr[Int((unwindAngle(bearing) / 45).rounded(.toNearestOrAwayFromZero)) % 8] }
  /// الفرق الموقّع بين اتجاهين (-180..180]: موجب = الهدف على يمين المرجع
  public static func signedDifference(target: Double, reference: Double) -> Double { quadrantShiftAngle(unwindAngle(target - reference)) }

  public struct SunMoments: Sendable { public let bearing: Double; public let sunAtQibla: Date?; public let shadowAtQibla: Date? }
  /// لحظات التحقق بالشمس خلال يوم مدني: الشمس في اتجاه القبلة، أو معاكسة لها فيشير الظل إلى القبلة
  public static func sunQiblaMoments(latitude: Double, longitude: Double, civil: CivilDate, tz: TimeZone) -> SunMoments {
    let bearing = info(latitude: latitude, longitude: longitude).bearing
    let start = civil.localMidnight(in: tz)
    func find(_ target: Double) -> Date? {
      var prev: Double? = nil; var prevT: Date? = nil
      var m = 0
      while m <= 1440 {
        let t = start.addingTimeInterval(Double(m) * 60)
        let p = Astro.sunPosition(at: t, latitude: latitude, longitude: longitude)
        let f = signedDifference(target: p.azimuth, reference: target)
        if let pv = prev, let pt = prevT, (f.sign != pv.sign) && f != 0 && pv != 0, abs(f - pv) < 180 {
          var lo = pt, hi = t, flo = pv
          for _ in 0..<25 {
            let mid = Date(timeIntervalSince1970: (lo.timeIntervalSince1970 + hi.timeIntervalSince1970) / 2)
            let fm = signedDifference(target: Astro.sunPosition(at: mid, latitude: latitude, longitude: longitude).azimuth, reference: target)
            if (fm.sign == flo.sign) { lo = mid; flo = fm } else { hi = mid }
          }
          let res = Date(timeIntervalSince1970: ((lo.timeIntervalSince1970 + hi.timeIntervalSince1970) / 2).rounded())
          let pr = Astro.sunPosition(at: res, latitude: latitude, longitude: longitude)
          if pr.altitude > 0 && pr.altitude < 85 && abs(signedDifference(target: pr.azimuth, reference: target)) < 0.5 { return res }
        }
        prev = f; prevT = t
        m += 2
      }
      return nil
    }
    return SunMoments(bearing: bearing, sunAtQibla: find(bearing), shadowAtQibla: find(unwindAngle(bearing + 180)))
  }

  public struct ZenithEvent: Sendable { public let time: Date; public let altitude: Double }
  /// حادثتا تعامد الشمس على الكعبة في سنة معيّنة (نحو 27/28 مايو و15/16 يوليو)
  public static func kaabaZenithEvents(year: Int) -> [ZenithEvent] {
    var events: [ZenithEvent] = []
    for (m1, d1, m2, d2) in [(5, 20, 6, 5), (7, 8, 7, 24)] {
      var best: ZenithEvent? = nil
      var jd = Astro.julianDay(year: year, month: m1, day: d1)
      let end = Astro.julianDay(year: year, month: m2, day: d2)
      while jd <= end {
        let civil = CivilDate(Date(timeIntervalSince1970: (jd - 2440587.5) * 86400), in: TimeZone(identifier: "UTC")!)
        let st = SolarTime(date: civil, coords: kaaba)
        let t = civil.utcMidnight.addingTimeInterval(st.transit * 3600)
        let alt = Astro.sunPosition(at: t, latitude: kaaba.latitude, longitude: kaaba.longitude).altitude
        if best == nil || alt > best!.altitude { best = ZenithEvent(time: t, altitude: alt) }
        jd += 1
      }
      if let best { events.append(best) }
    }
    return events
  }

  /// نقاط على قوس الدائرة العظمى من نقطة إلى أخرى — للرسم على الخريطة؛ n+1 نقطة [lat, lon]
  public static func greatCirclePoints(lat1: Double, lon1: Double, lat2: Double, lon2: Double, n: Int = 64) -> [(Double, Double)] {
    let d = Double.pi / 180
    func toVec(_ la: Double, _ lo: Double) -> (Double, Double, Double) { (cos(la * d) * cos(lo * d), cos(la * d) * sin(lo * d), sin(la * d)) }
    let a = toVec(lat1, lon1), b = toVec(lat2, lon2)
    let dot = max(-1, min(1, a.0 * b.0 + a.1 * b.1 + a.2 * b.2))
    let omega = acos(dot)
    if omega < 1e-7 { return [(lat1, lon1), (lat2, lon2)] }
    var pts: [(Double, Double)] = []
    for i in 0...n {
      let t = Double(i) / Double(n); let s1 = sin((1 - t) * omega) / sin(omega), s2 = sin(t * omega) / sin(omega)
      let x = s1 * a.0 + s2 * b.0, y = s1 * a.1 + s2 * b.1, z = s1 * a.2 + s2 * b.2
      pts.append((atan2(z, hypot(x, y)) / d, atan2(y, x) / d))
    }
    return pts
  }
}
